File: valo_tui/screens/widgets.py
"""Shared widgets and rendering helpers."""

from __future__ import annotations

from textual.app import ComposeResult
from textual.binding import Binding
from textual.widgets import DataTable, Label, Static

from ..data.models import MatchCard

LIVE = "#e8674e"
ACCENT = "#e8674e"
MUTED = "#4a708b"
TEXT = "#c8d8e8"
RULE = "#1c3a52"

BRAND = "valo-tui · vct26"

# Global nav: (key, route, label). Always visible.
GLOBAL_NAV = [
    ("h", "home", "home"),
    ("e", "events", "events"),
    ("l", "live", "live"),
    ("a", "about", "about"),
]

# Event-context nav: revealed only while an event is in focus. These are the
# children of an event, never global pages.
EVENT_NAV = [
    ("o", "overview", "overview"),
    ("r", "results", "results"),
    ("f", "fixtures", "fixtures"),
    ("t", "standings", "standings"),
    ("b", "bracket", "bracket"),
    ("m", "teams", "teams"),
]


class Sidebar(Static):
    """Focusable, context-aware nav rail.

    Shows the global section always; once an event is selected the app reveals
    that event's sub-pages beneath it. Up/Down move the highlight (switching the
    page live); Enter/Right enters the content.
    """

    can_focus = True

    BINDINGS = [
        Binding("up,k", "nav(-1)", "Up", show=False),
        Binding("down,j", "nav(1)", "Down", show=False),
        Binding("enter,right,l", "enter", "Open", show=False),
    ]

    def __init__(self, active: str = "home", **kwargs) -> None:
        self._active = active
        self._focused = False
        super().__init__("", **kwargs)

    def on_mount(self) -> None:
        self.update(self._markup())

    def _nav(self) -> list[tuple[str, str, str]]:
        """The currently selectable items (global, plus event items if any)."""
        items = list(GLOBAL_NAV)
        if getattr(self.app, "current_event_id", None) is not None:
            items += EVENT_NAV
        return items

    def _index(self) -> int:
        for i, (_, route, _) in enumerate(self._nav()):
            if route == self._active:
                return i
        return 0

    @property
    def _route(self) -> str:
        return self._nav()[self._index()][1]

    def set_active(self, route: str) -> None:
        self._active = route
        self.update(self._markup())

    def rebuild(self) -> None:
        """Re-render after the event context changed."""
        self.update(self._markup())

    def action_nav(self, delta: int) -> None:
        nav = self._nav()
        self._active = nav[(self._index() + delta) % len(nav)][1]
        self.update(self._markup())
        self.app.switch_content(self._active)

    def action_enter(self) -> None:
        self.app.switch_content(self._route)
        self.app.focus_content()

    def _markup(self) -> str:
        active = self._active
        focused = self._focused

        def row(key: str, route: str, label: str) -> str:
            if route == active:
                marker = f"[{ACCENT}]›[/]" if focused else " "
                return f"{marker}[{MUTED}][{key}][/] [bold {ACCENT}]{label}[/]"
            return f" [{MUTED}][{key}][/] [{TEXT}]{label}[/]"

        lines = [f"[bold {TEXT}]{BRAND}[/]", f"[{RULE}]{'─' * 20}[/]", ""]
        lines += [row(k, r, lbl) for k, r, lbl in GLOBAL_NAV]

        if getattr(self.app, "current_event_id", None) is not None:
            name = _clip(getattr(self.app, "event_name", "") or "event", 18)
            lines += [
                "",
                f"[{MUTED}]── event ──[/]",
                f"[bold {ACCENT}]{name}[/]",
                "",
            ]
            lines += [row(k, r, lbl) for k, r, lbl in EVENT_NAV]

        lines += [
            "",
            f"[{RULE}]{'─' * 20}[/]",
            "",
            f"[{MUTED}]↑↓    navigate[/]",
            f"[{MUTED}]enter open[/]",
            f"[{MUTED}]esc   back here[/]",
            f"[{MUTED}]q     quit[/]",
        ]
        return "\n".join(lines)

    def on_focus(self) -> None:
        self._focused = True
        self.update(self._markup())

    def on_blur(self) -> None:
        self._focused = False
        self.update(self._markup())


class VimDataTable(DataTable):
    """DataTable with vim-style j/k bound to its built-in cursor actions."""

    BINDINGS = [
        Binding("j", "cursor_down", "Down", show=False),
        Binding("k", "cursor_up", "Up", show=False),
    ]


class LiveDot(Static):
    """A live indicator that pulses ● ↔ ○ on a slow interval."""

    DEFAULT_CSS = "LiveDot { width: auto; color: #e87a5d; }"

    def __init__(self, label: str = "LIVE", **kwargs) -> None:
        self._label = label
        self._on = True
        super().__init__(self._markup(), **kwargs)

    def _markup(self) -> str:
        dot = "●" if self._on else "○"
        return f"[{LIVE}]{dot} {self._label}[/]" if self._label else f"[{LIVE}]{dot}[/]"

    def on_mount(self) -> None:
        self.set_interval(0.7, self._tick)

    def _tick(self) -> None:
        self._on = not self._on
        self.update(self._markup())


class SkeletonRow(Static):
    """A shimmering placeholder row shown while real data is unavailable."""

    DEFAULT_CSS = "SkeletonRow { width: 1fr; height: 1; color: #22384a; }"
    _FRAMES = ["░░░░", "▒▒▒▒", "▓▓▓▓", "▒▒▒▒"]

    def __init__(self, cells: int = 8, **kwargs) -> None:
        self._cells = cells
        self._i = 0
        super().__init__(self._frame(), **kwargs)

    def _frame(self) -> str:
        return "   ".join([self._FRAMES[self._i]] * self._cells)

    def on_mount(self) -> None:
        self.set_interval(0.18, self._tick)

    def _tick(self) -> None:
        self._i = (self._i + 1) % len(self._FRAMES)
        self.update(self._frame())


def match_line(m: MatchCard) -> str:
    """One-line Rich-markup summary of a match for the dashboard panels."""
    if m.is_live:
        dot = f"[{LIVE}]●[/] "
        score = f"[{LIVE}]{_score(m)}[/]"
    elif m.status == "completed":
        dot = f"[{MUTED}]·[/] "
        score = f"[{MUTED}]{_score(m)}[/]"
    else:
        dot = f"[{MUTED}]○[/] "
        score = f"[{MUTED}]{m.time or 'soon'}[/]"
    t1 = _clip(m.team1.name, 12)
    t2 = _clip(m.team2.name, 12)
    return f"{dot}{t1} [{MUTED}]vs[/] {t2}  {score}"


def _score(m: MatchCard) -> str:
    s1 = m.team1.score if m.team1.score is not None else "–"
    s2 = m.team2.score if m.team2.score is not None else "–"
    return f"{s1}–{s2}"


def _clip(s: str, n: int) -> str:
    return s if len(s) <= n else s[: n - 1] + "…"

File: internal/widgets/widgets.go
// Package widgets holds reusable render helpers: the context-aware sidebar and
// the one-line match summary. Ported from valo_tui/screens/widgets.py.
package widgets

import (
	"strings"

	"charm.land/lipgloss/v2"

	"github.com/jashkarangiya/valo-tui/internal/data"
	"github.com/jashkarangiya/valo-tui/internal/styles"
)

const brand = "valo-tui · vct26"

// NavItem is one selectable rail entry: hotkey, route id, label.
type NavItem struct {
	Key, Route, Label string
}

// GlobalNav is always visible.
var GlobalNav = []NavItem{
	{"h", "home", "home"},
	{"e", "events", "events"},
	{"l", "live", "live"},
	{"a", "about", "about"},
}

// EventNav is revealed only while an event is in focus.
var EventNav = []NavItem{
	{"o", "overview", "overview"},
	{"r", "results", "results"},
	{"f", "fixtures", "fixtures"},
	{"t", "standings", "standings"},
	{"b", "bracket", "bracket"},
	{"m", "teams", "teams"},
}

var (
	mutedS  = lipgloss.NewStyle().Foreground(styles.Muted)
	textS   = lipgloss.NewStyle().Foreground(styles.Text)
	accentS = lipgloss.NewStyle().Foreground(styles.Accent)
	accentB = lipgloss.NewStyle().Foreground(styles.Accent).Bold(true)
	textB   = lipgloss.NewStyle().Foreground(styles.Text).Bold(true)
	ruleS   = lipgloss.NewStyle().Foreground(styles.Rule)
	liveS   = lipgloss.NewStyle().Foreground(styles.Live)
	rowSel  = lipgloss.NewStyle().Foreground(styles.Accent).Background(styles.SelBg).Bold(true)
)

// navRow renders one rail entry. The active page ALWAYS shows the › cursor and
// a highlighted [key] + label so you can see where you are; when the rail holds
// keyboard focus the active row also gets a background highlight.
func navRow(it NavItem, active, focused bool) string {
	if !active {
		return "  " + mutedS.Render("["+it.Key+"]") + " " + textS.Render(it.Label)
	}
	if focused {
		return rowSel.Render("› [" + it.Key + "] " + it.Label)
	}
	return accentS.Render("› ["+it.Key+"]") + " " + accentB.Render(it.Label)
}

// buildSidebar is the single source of the rail layout: it returns the rendered
// lines and a map from line index → route for click hit-testing.
func buildSidebar(active, eventName string, focused bool) ([]string, map[int]string) {
	var lines []string
	routes := map[int]string{}
	add := func(s string) { lines = append(lines, s) }
	addNav := func(it NavItem) {
		routes[len(lines)] = it.Route
		add(navRow(it, it.Route == active, focused))
	}

	add(textB.Render(brand))
	add(ruleS.Render(strings.Repeat("─", 20)))
	add("")
	for _, it := range GlobalNav {
		addNav(it)
	}
	if eventName != "" {
		add("")
		add(mutedS.Render("── event ──"))
		add(accentB.Render(clip(eventName, 18)))
		add("")
		for _, it := range EventNav {
			addNav(it)
		}
	}
	add("")
	add(ruleS.Render(strings.Repeat("─", 20)))
	add("")
	add(mutedS.Render("↑↓    navigate"))
	add(mutedS.Render("enter  open"))
	add(mutedS.Render("esc    back to rail"))
	add(mutedS.Render("q      quit"))
	return lines, routes
}

// Sidebar renders the nav rail for the current route.
func Sidebar(active, eventName string, focused bool) string {
	lines, _ := buildSidebar(active, eventName, focused)
	return strings.Join(lines, "\n")
}

// SidebarRoutes maps a sidebar text-line index to the route on that line (for
// mouse clicks). Only depends on whether an event is in focus.
func SidebarRoutes(eventName string) map[int]string {
	_, routes := buildSidebar("", eventName, false)
	return routes
}

// MatchLine is a one-line summary of a match for the dashboard panels.
func MatchLine(m data.MatchCard) string {
	var dot, score string
	switch {
	case m.IsLive():
		dot = liveS.Render("●") + " "
		score = liveS.Render(scoreOf(m))
	case m.Status == "completed":
		dot = mutedS.Render("·") + " "
		score = mutedS.Render(scoreOf(m))
	default:
		dot = mutedS.Render("○") + " "
		t := m.Time
		if t == "" {
			t = "soon"
		}
		score = mutedS.Render(t)
	}
	t1 := clip(m.Team1.Name, 12)
	t2 := clip(m.Team2.Name, 12)
	return dot + t1 + " " + mutedS.Render("vs") + " " + t2 + "  " + score
}

func scoreOf(m data.MatchCard) string {
	s1, s2 := "–", "–"
	if m.Team1.Score != nil {
		s1 = itoa(*m.Team1.Score)
	}
	if m.Team2.Score != nil {
		s2 = itoa(*m.Team2.Score)
	}
	return s1 + "–" + s2
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

func clip(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n-1]) + "…"
}
